iceoryx_posh/source/roudi/service_registry.cpp

```diff
@@ -16,45 +17,164 @@
 
 #include "iceoryx_posh/internal/roudi/service_registry.hpp"
 
+#include <algorithm>
+
 namespace iox
 {
 namespace roudi
 {
-void ServiceRegistry::add(const CaproIdString_t& service, const CaproIdString_t& instance)
+cxx::expected<ServiceRegistry::ServiceRegistryError>
+ServiceRegistry::add(const capro::ServiceDescription& serviceDescription)
 {
-    cxx::set::add(m_serviceMap[service].instanceSet, instance);
+    // Forbid duplicate service descriptions entries
+    for (auto& element : m_serviceDescriptionVector)
+    {
+        if (element == serviceDescription)
+        {
+            return cxx::error<ServiceRegistryError>(ServiceRegistryError::SERVICE_DESCRIPTION_ALREADY_ADDED);
```

Comment:
This error just increases the error handling overhead and has no value for the user. If a user calls `add`, there are two possibilities:
1. it is not added, but after the call -> it is added
2. it is added, and after the call -> it is still added

Therefore, the user does not require any action here and if they would like to know if a certain service was registered they always can use `find`.

Reply:
Yes, it is also not an error if it already exists. Whether the user should be informed is debatable. I am fine with either way, just do not classify it as error.